### PR TITLE
Add support for template versioning in the application management service

### PR DESCRIPTION
**Pull request by DilshanSenarath**
### Proposed changes in this pull request

- Add support for specifying the template version when creating an application
- Add support for updating the template version of an application
- Add support for retrieving the `templateVersion` when fetching the specified service provider data
- Add support for requesting the `templateVersion` as a `requiredAttribute` in application listing data

### Related Issues

- https://github.com/wso2/product-is/issues/20108

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/ServiceProvider.java [hunk 1]
```diff
@@ -131,6 +132,10 @@ public class ServiceProvider implements Serializable {
     @XmlElement(name = TEMPLATE_ID)
     private String templateId;
 
+    @IgnoreNullElement
+    @XmlElement(name = TEMPLATE_VERSION)
+    private String templateVersion;
```
Comment (ImalshaG) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/ServiceProvider.java:137]: Can you give an example of what type of values the `templateVersion` would hold?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/test/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImplTest.java [hunk 2]
```diff
@@ -1136,6 +1143,60 @@ public void testGetTrustedApps(SpTrustedAppMetadata trustedAppMetadata1, SpTrust
         applicationManagementService.deleteApplications(SUPER_TENANT_ID);
     }
 
+    @DataProvider(name = "testAddApplicationWithTemplateIdAndTemplateVersion")
+    public Object[][] testAddApplicationWithTemplateIdAndTemplateVersion() {
```
Comment (ImalshaG) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/test/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImplTest.java:1147]:
```suggestion
    public Object[][] addApplicationWithTemplateIdAndTemplateVersionData() {
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/test/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImplTest.java [hunk 3]
```diff
@@ -1136,6 +1143,60 @@ public void testGetTrustedApps(SpTrustedAppMetadata trustedAppMetadata1, SpTrust
...
+
+        return new Object[][] {
+                {APPLICATION_TEMPLATE_ID_1, APPLICATION_TEMPLATE_VERSION_1}
```
Comment (ImalshaG) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/test/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImplTest.java:1150]: Let's add a test case with null values as well

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/ServiceProvider.java (hunk 1 above)
Comment (DilshanSenarath) [reply components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/ServiceProvider.java:137]: It can hold values like v1.0.0

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/test/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImplTest.java (hunk 2 above)
Comment (DilshanSenarath) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/test/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImplTest.java]: https://github.com/wso2/carbon-identity-framework/pull/5801/commits/ad555d1e9871a12480c146f2eee3b8123a3126f3

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/test/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImplTest.java [hunk 4]
```diff
@@ -1136,6 +1143,60 @@ public void testGetTrustedApps(SpTrustedAppMetadata trustedAppMetadata1, SpTrust
...
+        };
+    }
+
+    @Test(dataProvider = "testAddApplicationWithTemplateIdAndTemplateVersion")
+    public void testAddApplicationWithTemplateIdAndTemplateVersion(String templateId, String templateVersion)
+            throws Exception {
+        ServiceProvider inputSP = new ServiceProvider();
```
[the remark this answers is not in the document]
Comment (DilshanSenarath) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/test/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImplTest.java:1158]: https://github.com/wso2/carbon-identity-framework/pull/5801/commits/3bb30d8cedd9b55f1c14e6791d48cc4db3b497cc

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/test/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImplTest.java (hunk 3 above)
Comment (DilshanSenarath) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/test/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImplTest.java]: https://github.com/wso2/carbon-identity-framework/pull/5801/commits/098d5fa80c4cc5593886d4ac16a3b849c6777fc5

**Review by ImalshaG — APPROVED** [review]

**Review by ImalshaG — APPROVED** [review]
